cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/LineageAdmin.java

```diff
@@ -104,145 +107,123 @@ public Lineage computeLineage(final DatasetId sourceDataset, long startMillis, l
+          // determine if a dataset is local dataset. The local dataset always ends with the workflow run id
+          if (rollUpWorkflow) {
+            // Step 1 walk through the program relations, filter out the possible mapreduce and spark programs that
+            // could be in the workflow, and get the appSpec for the program, to determine what other programs
+            // are in the workflow
+            Map<ApplicationId, ApplicationSpecification> appSpecs = new HashMap<>();
+            Set<ProgramRunId> possibleInnerPrograms = new HashSet<>();
+            programRelations.forEach(relation -> {
+              ProgramType type = relation.getProgram().getType();
+              if (type.equals(ProgramType.MAPREDUCE) || type.equals(ProgramType.SPARK)) {
+                possibleInnerPrograms.add(relation.getProgramRunId());
+                ApplicationId appId = relation.getProgram().getParent();
+                if (!appSpecs.containsKey(appId)) {
+                  appSpecs.put(appId, store.getApplication(appId));
+                }
+              }
+            });
+
+            // Step 2, get the run record for all the possible inner programs, the run record contains the
+            // workflow information, fetch the workflow id and add them to the map
+            Map<ProgramRunId, RunRecordMeta> runRecords = store.getRuns(possibleInnerPrograms);
+            Set<ProgramRunId> workflowRunIds = new HashSet<>();
+            runRecords.forEach((programRunId, runRecord) -> {
+              // the run record can be null if a program run is not found, but this should not happen since we
+              // query by the program run id
+              if (runRecord != null && runRecord.getSystemArgs().containsKey(ProgramOptionConstants.WORKFLOW_RUN_ID)) {
+                Map<String, String> systemArgs = runRecord.getSystemArgs();
+                String workflowRunId = systemArgs.get(ProgramOptionConstants.WORKFLOW_RUN_ID);
+                String workflowName = systemArgs.get(ProgramOptionConstants.WORKFLOW_NAME);
+                WorkflowId workflowId = programRunId.getParent().getParent().workflow(workflowName);
+                ProgramRunId wfRunId = workflowId.run(workflowRunId);
+                programWorkflowMap.put(programRunId, wfRunId);
+                workflowRunIds.add(wfRunId);
+              }
+            });
+
+            // Step 3, fetch run records of the workflow, the properties of the workflow run record has all
+            // the inner program run ids, compare them with the app spec to get the type of the program
+            runRecords = store.getRuns(workflowRunIds);
+            runRecords.forEach((programRunId, runRecord) -> {
+              if (runRecord != null) {
+                runRecord.getProperties().forEach((key, value) -> {
+                  ApplicationId appId = programRunId.getParent().getParent();
```

Comment:
This and the next 2 line seems should be outside of the `forEach` as they are loop invariants.

Reply:
This app id is generated from the programRunId, which is the key of the map